### PR TITLE
Result variable for DSL

### PR DESCRIPTION
**Pull request by sheepdreamofandroids**
This adds an extra parameter ("variable") to the DSL step. It is used as the name of an environment variable that will receive the comma separated list of the names of the locked resources while the block executes.

### CODE REVIEW

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 1]
```diff
@@ -62,10 +66,27 @@ public static void proceed(List<String> resourcenames, StepContext context, Stri
...
+		try {
+			BodyInvoker bodyInvoker = context.newBodyInvoker().
+				withCallback(new Callback(resourcenames, resourceDescription, inversePrecedence)).
+				withDisplayName(null);
+			if(variable != null && variable.length()>0)
+				// set the variable for the duration of the block
+				bodyInvoker.withContext(EnvironmentExpander.merge(context.get(EnvironmentExpander.class), new EnvironmentExpander() {
+					@Override
+					public void expand(EnvVars env) throws IOException, InterruptedException {
+						final String resources = resourcenames.toString();
```
Comment (abishek3876) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: Iterating the list would be both efficient and consistent than this. List#toString doesn't even define any kind of contract.

Comment (sheepdreamofandroids) [reply src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: You're right, I was too lazy...

**Comment by bk00041** [comment]
+1 for this feature. 
Any plans to implement soon?

**Comment by scott-zhong** [comment]
+1 on this.
We are manually building the plugin with this merge request.  When can we expect this merge to be accepted?

**Comment by jordburns** [comment]
+ 1
Can anyone provide the hpi for this, we are struggling to build this internally.  please merge

**Comment by amuniz** [comment]
Merged. Delaying the release a bit until I review other PRs that could be released together.

**Comment by meuter** [comment]
Small question to those of you who use a recompiled version of the plugin including this PR: I have 6 instances of a particular resource type available in Jenkins and a pipeline with a test stage including 10 parallel branches. Each branch requests one instance of the resource before starting.

What I'm observing is that 6 branches start and receive one instance properly through the environment variable. However, for the other 4 branches, instead of blocking and waiting for some resource instances to become available, they start immediately and the environment variable that is supposed to contain the resource instance is not set. 

In the Jenkins file, I'm using the following piece of code to acquire the resource:
`lock(label: "test-server", quantity:1, resource:null, variable:"SERVER_IP") { ... }`

Anyone else is encountering this?

**Comment by anthuwan13** [comment]
Should I update the lockable resource plugging to get this latest change work in my jenkins ?

**Comment by boblloyd** [comment]
I'm finding that the variable is not set if the lock has to queue waiting on a resource to be available.

```groovy
def steps = [:]

steps['stage1'] = {
    lock(label: 'resourcePool', quantity: 1, variable: "STAGE1_VM"){
        echo "locked resource: ${env.STAGE1_VM}"
        echo sh(returnStdout: true, script: 'env')
        sleep 10
    }
}

steps['stage2'] = {
    lock(label: 'resourcePool', quantity: 1, variable: "STAGE2_VM"){
        echo "locked resource: ${env.STAGE2_VM}"
        echo sh(returnStdout: true, script: 'env')
        sleep 15
    }
}
steps['stage3'] = {
    lock(label: 'resourcePool', quantity: 1, variable: "STAGE3_VM"){
        echo "locked resource: ${env.STAGE3_VM}"
        echo sh(returnStdout: true, script: 'env')
        sleep 5
    }
}
steps['stage4'] = {
    lock(label: 'resourcePool', quantity: 1, variable: "STAGE4_VM"){
        echo "locked resource: ${env.STAGE4_VM}"
        echo sh(returnStdout: true, script: 'env')
        sleep 20
    }
}

node('master'){
    parallel steps
}
```

Results in stage3 and stage4 printing 'null' for the env.STAGE3_VM and env.STAGE4_VM variables, even though resources are locked correctly.

**Comment by amuniz** [comment]
@boblloyd 

> Results in stage3 and stage4 printing 'null' for the env.STAGE3_VM and env.STAGE4_VM variables, even though resources are locked correctly.

Interesting, can you file a bug in JIRA please? Just copy & paste your comment in the description. Thanks.

Maybe @sheepdreamofandroids can have a look as the author of the feature? I'll do it otherwise.

**Comment by sheepdreamofandroids** [comment]
I see only one way that could happen: the parameter `final List<String> resourcenames` of the `proceed` method is modified before the block starts.
The easiest way to fix that is to move the line `final String resources = COMMA_JOINER.join(resourcenames);` out of the inner class to right after the enclosing `if` statement.

**Pull request merged**